File: src/features/library/api/library.ts
import { invoke, isTauri } from "@tauri-apps/api/core";
import { open } from "@tauri-apps/plugin-dialog";
import type {
  LibraryAnnotation,
  LibraryAnnotationCreate,
  LibraryAnnotationUpdate,
  LibraryCollection,
  LibraryImportResult,
  LibraryItem,
  LibraryItemUpdate,
  LibraryListPage,
  LibraryListRequest,
  LibraryNote,
  LibraryNoteImportResult,
  LibraryNoteCreate,
  LibraryNoteSummary,
  LibraryNoteUpdate,
  LibraryReadingState,
  LibraryReadingStateUpdate,
} from "../types";

export function isLibraryRuntime() {
  return isTauri();
}

export async function chooseLibraryPdfFiles(): Promise<string[]> {
  if (!isTauri()) return [];
  const selected = await open({
    title: "导入 PDF 文献",
    multiple: true,
    directory: false,
    filters: [{ name: "PDF 文献", extensions: ["pdf"] }],
  });
  if (typeof selected === "string") return [selected];
  return selected ?? [];
}

export async function chooseLibraryMarkdownFiles(): Promise<string[]> {
  if (!isTauri()) return [];
  const selected = await open({
    title: "导入 Markdown 笔记",
    multiple: true,
    directory: false,
    filters: [{ name: "Markdown 笔记", extensions: ["md", "markdown"] }],
  });
  if (typeof selected === "string") return [selected];
  return selected ?? [];
}

export function listLibraryItems(request: LibraryListRequest) {
  if (!isTauri()) {
    return Promise.resolve<LibraryListPage>({ items: [], offset: 0, total: 0, hasMore: false });
  }
  return invoke<LibraryListPage>("library_list_items", { request });
}

export function getLibraryItem(itemId: string) {
  return invoke<LibraryItem>("library_get_item", { itemId });
}

export function importLibraryPdfs(paths: string[], collectionId: string | null) {
  return invoke<LibraryImportResult>("library_import_pdfs", { paths, collectionId });
}

export function updateLibraryItem(update: LibraryItemUpdate) {
  return invoke<LibraryItem>("library_update_item", { update });
}

export function setLibraryItemFavorite(itemId: string, favorite: boolean) {
  return invoke<LibraryItem>("library_set_favorite", { itemId, favorite });
}

export function moveLibraryItemToTrash(itemId: string) {
  return invoke<LibraryItem>("library_move_to_trash", { itemId });
}

export function restoreLibraryItem(itemId: string) {
  return invoke<LibraryItem>("library_restore_item", { itemId });
}

export function deleteLibraryItemPermanently(itemId: string) {
  return invoke<boolean>("library_delete_permanently", { itemId });
}

export function markLibraryItemOpened(itemId: string) {
  return invoke<LibraryItem>("library_mark_opened", { itemId });
}

export function openLibraryItem(itemId: string) {
  return invoke<LibraryItem>("library_open_item", { itemId });
}

export async function readLibraryPdfRange(
  itemId: string,
  start: number,
  end: number,
): Promise<Uint8Array> {
  if (!isTauri()) throw new Error("PDF 阅读器只能在桌面应用中使用。");
  const raw = await invoke<ArrayBuffer | Uint8Array | number[]>("library_read_pdf_range", {
    itemId,
    start,
    end,
  });
  if (raw instanceof Uint8Array) return raw;
  if (raw instanceof ArrayBuffer) return new Uint8Array(raw);
  return Uint8Array.from(raw);
}

export function getLibraryReadingState(itemId: string) {
  return invoke<LibraryReadingState>("library_get_reading_state", { itemId });
}

export function saveLibraryReadingState(update: LibraryReadingStateUpdate) {
  return invoke<LibraryReadingState>("library_save_reading_state", { update });
}

export function listLibraryAnnotations(itemId: string) {
  if (!isTauri()) return Promise.resolve<LibraryAnnotation[]>([]);
  return invoke<LibraryAnnotation[]>("library_list_annotations", { itemId });
}

export function createLibraryAnnotation(create: LibraryAnnotationCreate) {
  return invoke<LibraryAnnotation>("library_create_annotation", { create });
}

export function updateLibraryAnnotation(update: LibraryAnnotationUpdate) {
  return invoke<LibraryAnnotation>("library_update_annotation", { update });
}

export function deleteLibraryAnnotation(annotationId: string) {
  return invoke<boolean>("library_delete_annotation", { annotationId });
}

export function listLibraryNotes(itemId?: string | null) {
  if (!isTauri()) return Promise.resolve<LibraryNoteSummary[]>([]);
  return invoke<LibraryNoteSummary[]>("library_list_notes", { itemId: itemId ?? null });
}

export function getLibraryNote(noteId: string) {
  return invoke<LibraryNote>("library_get_note", { noteId });
}

export function createLibraryNote(create: LibraryNoteCreate) {
  return invoke<LibraryNote>("library_create_note", { create });
}

export function importLibraryMarkdownNotes(paths: string[]) {
  return invoke<LibraryNoteImportResult>("library_import_markdown_notes", { paths });
}

export function updateLibraryNote(update: LibraryNoteUpdate) {
  return invoke<LibraryNote>("library_update_note", { update });
}

export function deleteLibraryNote(noteId: string) {
  return invoke<boolean>("library_delete_note", { noteId });
}

export function listLibraryCollections() {
  if (!isTauri()) return Promise.resolve<LibraryCollection[]>([]);
  return invoke<LibraryCollection[]>("library_list_collections");
}

export function createLibraryCollection(name: string) {
  return invoke<LibraryCollection>("library_create_collection", { name });
}

export function renameLibraryCollection(collectionId: string, name: string) {
  return invoke<void>("library_rename_collection", { collectionId, name });
}

export function deleteLibraryCollection(collectionId: string) {
  return invoke<boolean>("library_delete_collection", { collectionId });
}

File: src-tauri/src/commands/library.rs
//! Work 文献库 Tauri 命令边界。
//!
//! SQLite、哈希和文件复制都放入阻塞线程；所有写操作通过同一异步锁串行化。

use tauri::ipc::Response;
use tauri::{AppHandle, State};
use tauri_plugin_opener::OpenerExt;

use crate::{
    library::types::{
        LibraryAnnotation, LibraryAnnotationCreate, LibraryAnnotationUpdate, LibraryCollection,
        LibraryImportResult, LibraryItem, LibraryItemUpdate, LibraryListPage, LibraryListRequest,
        LibraryNote, LibraryNoteCreate, LibraryNoteImportResult, LibraryNoteSummary, LibraryNoteUpdate, LibraryReadingState,
        LibraryReadingStateUpdate,
    },
    state::AppState,
};

fn join_error(error: impl std::fmt::Display) -> String {
    format!("文献库后台任务失败：{error}")
}

#[tauri::command]
pub async fn library_list_items(
    state: State<'_, AppState>,
    request: LibraryListRequest,
) -> Result<LibraryListPage, String> {
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.list_items(request))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_get_item(
    state: State<'_, AppState>,
    item_id: String,
) -> Result<LibraryItem, String> {
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.get_item(&item_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_import_pdfs(
    state: State<'_, AppState>,
    paths: Vec<String>,
    collection_id: Option<String>,
) -> Result<LibraryImportResult, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.import_pdfs(paths, collection_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_update_item(
    state: State<'_, AppState>,
    update: LibraryItemUpdate,
) -> Result<LibraryItem, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.update_item(update))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_set_favorite(
    state: State<'_, AppState>,
    item_id: String,
    favorite: bool,
) -> Result<LibraryItem, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.set_favorite(&item_id, favorite))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_move_to_trash(
    state: State<'_, AppState>,
    item_id: String,
) -> Result<LibraryItem, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.move_to_trash(&item_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_restore_item(
    state: State<'_, AppState>,
    item_id: String,
) -> Result<LibraryItem, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.restore_from_trash(&item_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_delete_permanently(
    state: State<'_, AppState>,
    item_id: String,
) -> Result<bool, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.delete_permanently(&item_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_mark_opened(
    state: State<'_, AppState>,
    item_id: String,
) -> Result<LibraryItem, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.mark_opened(&item_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_open_item(
    app: AppHandle,
    state: State<'_, AppState>,
    item_id: String,
) -> Result<LibraryItem, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    let (item, path) = tauri::async_runtime::spawn_blocking(move || {
        let item = repository.mark_opened(&item_id)?;
        let path = repository.primary_file_path(&item_id)?;
        Ok::<_, String>((item, path))
    })
    .await
    .map_err(join_error)??;
    app.opener()
        .open_path(path.to_string_lossy().into_owned(), None::<String>)
        .map_err(|error| format!("使用系统阅读器打开 PDF 失败：{error}"))?;
    Ok(item)
}

#[tauri::command]
pub async fn library_read_pdf_range(
    state: State<'_, AppState>,
    item_id: String,
    start: u64,
    end: u64,
) -> Result<Response, String> {
    let repository = state.library_repository.clone();
    let bytes = tauri::async_runtime::spawn_blocking(move || {
        repository.read_pdf_range(&item_id, start, end)
    })
    .await
    .map_err(join_error)??;
    Ok(Response::new(bytes))
}

#[tauri::command]
pub async fn library_get_reading_state(
    state: State<'_, AppState>,
    item_id: String,
) -> Result<LibraryReadingState, String> {
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.get_reading_state(&item_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_save_reading_state(
    state: State<'_, AppState>,
    update: LibraryReadingStateUpdate,
) -> Result<LibraryReadingState, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.save_reading_state(update))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_list_annotations(
    state: State<'_, AppState>,
    item_id: String,
) -> Result<Vec<LibraryAnnotation>, String> {
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.list_annotations(&item_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_create_annotation(
    state: State<'_, AppState>,
    create: LibraryAnnotationCreate,
) -> Result<LibraryAnnotation, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.create_annotation(create))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_update_annotation(
    state: State<'_, AppState>,
    update: LibraryAnnotationUpdate,
) -> Result<LibraryAnnotation, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.update_annotation(update))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_delete_annotation(
    state: State<'_, AppState>,
    annotation_id: String,
) -> Result<bool, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.delete_annotation(&annotation_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_list_notes(
    state: State<'_, AppState>,
    item_id: Option<String>,
) -> Result<Vec<LibraryNoteSummary>, String> {
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.list_notes(item_id.as_deref()))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_get_note(
    state: State<'_, AppState>,
    note_id: String,
) -> Result<LibraryNote, String> {
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.get_note(&note_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_create_note(
    state: State<'_, AppState>,
    create: LibraryNoteCreate,
) -> Result<LibraryNote, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.create_note(create))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_import_markdown_notes(
    state: State<'_, AppState>,
    paths: Vec<String>,
) -> Result<LibraryNoteImportResult, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.import_markdown_notes(paths))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_update_note(
    state: State<'_, AppState>,
    update: LibraryNoteUpdate,
) -> Result<LibraryNote, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.update_note(update))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_delete_note(
    state: State<'_, AppState>,
    note_id: String,
) -> Result<bool, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.delete_note(&note_id))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_list_collections(
    state: State<'_, AppState>,
) -> Result<Vec<LibraryCollection>, String> {
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.list_collections())
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_create_collection(
    state: State<'_, AppState>,
    name: String,
) -> Result<LibraryCollection, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.create_collection(&name))
        .await
        .map_err(join_error)?
}

#[tauri::command]
pub async fn library_rename_collection(
    state: State<'_, AppState>,
    collection_id: String,
    name: String,
) -> Result<(), String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || {
        repository.rename_collection(&collection_id, &name)
    })
    .await
    .map_err(join_error)?
}

#[tauri::command]
pub async fn library_delete_collection(
    state: State<'_, AppState>,
    collection_id: String,
) -> Result<bool, String> {
    let _write_guard = state.library_operations.lock().await;
    let repository = state.library_repository.clone();
    tauri::async_runtime::spawn_blocking(move || repository.delete_collection(&collection_id))
        .await
        .map_err(join_error)?
}
